File: client/lib/src/directory/directory_record_verifier.dart
import 'dart:typed_data';

import 'package:sodium/sodium.dart';

import '../crypto/ed25519.dart';
import 'directory_record.dart';
import 'directory_record_codec.dart';
import 'directory_rejection.dart';
import 'root_quorum.dart';

/// Resolves a parent intermediate record by its key_id.
/// Returns the raw signed CBOR bytes, or throws [ParentNotFound].
typedef ParentLookup = Uint8List Function(Uint8List parentKeyId);

/// Verifies a [DirectoryRecord] against a [PinnedRoot].
///
/// Rejection order (mirrors Java):
/// 1. Parse / canonical check   → [MalformedCbor]
/// 2. Version check             → [UnsupportedVersion]
/// 3. record_type validation    → [InvalidRecordType]
/// 4. key_use validation        → [InvalidKeyUse]
/// 5. status validation         → [InvalidStatus]
/// 6. Freshness check           → [FreshnessExpired]
/// 7. Signature container XOR   → [SignatureContainerConflict] / [SignatureContainerMissing]
/// 8. Root quorum / parent chain check
class DirectoryRecordVerifier {
  static const Duration _maxAge = Duration(days: 7);
  static const int _maxDepth = 2;

  static const _validRecordTypes = {'verifier', 'issuer', 'intermediate'};
  static const _validKeyUses = {'enc', 'auth', 'sign'};
  static const _validStatuses = {'active', 'superseded', 'revoked'};

  final DirectoryRecordCodec _codec;
  final Sodium _sodium;
  final int Function() _nowMs;

  DirectoryRecordVerifier(this._sodium, {int Function()? nowMs})
      : _codec = DirectoryRecordCodec(),
        _nowMs = nowMs ?? (() => DateTime.now().millisecondsSinceEpoch);

  /// Verifies [recordBytes] against [pinned].
  ///
  /// Fails fast if a parent chain is required and no [lookup] is supplied.
  DirectoryRecord verify(Uint8List recordBytes, PinnedRoot pinned,
      [ParentLookup? lookup]) {
    return _verifyAtDepth(recordBytes, pinned, lookup, 1);
  }

  DirectoryRecord _verifyAtDepth(Uint8List recordBytes, PinnedRoot pinned,
      ParentLookup? lookup, int depth) {
    final record = _codec.decode(recordBytes); // may throw MalformedCbor / XOR

    if (record.version != 1) {
      throw UnsupportedVersion(record.version);
    }
    if (!_validRecordTypes.contains(record.recordType)) {
      throw InvalidRecordType(record.recordType);
    }
    if (!_validKeyUses.contains(record.keyUse)) {
      throw InvalidKeyUse(record.keyUse);
    }
    if (!_validStatuses.contains(record.status)) {
      throw InvalidStatus(record.status);
    }

    final nowMs = _nowMs();
    final ageMs = nowMs - record.issuedAt;
    if (ageMs > _maxAge.inMilliseconds) {
      throw FreshnessExpired(
          'Record issued_at=${record.issuedAt} is older than 7 days '
          '(age=${ageMs}ms)');
    }

    final signedBytes = _codec.signedBytesOfEncoded(recordBytes);

    if (record.parentKeyId == null) {
      RootQuorum.verify(
          _sodium, signedBytes, record.rootSignatures, record.issuedAt, pinned);
    } else {
      _verifyParentChain(record, signedBytes, pinned, lookup, depth);
    }

    return record;
  }

  void _verifyParentChain(DirectoryRecord record, Uint8List signedBytes,
      PinnedRoot pinned, ParentLookup? lookup, int depth) {
    if (depth >= _maxDepth) {
      throw ChainTooDeep('Chain depth exceeds maximum of $_maxDepth');
    }

    if (lookup == null) {
      throw ParentNotFound(
          'Parent lookup not available for key_id=${_hex(record.parentKeyId!)}');
    }

    final parentBytes = lookup(record.parentKeyId!);
    final parent = _codec.decode(parentBytes);

    if (parent.recordType != 'intermediate') {
      throw ParentNotIntermediate(
          "Parent record_type='${parent.recordType}' is not 'intermediate'");
    }
    if (parent.status != 'active') {
      throw ParentInactive("Parent status='${parent.status}' is not 'active'");
    }
    final now = _nowMs();
    if (now < parent.validFrom || now >= parent.validUntil) {
      throw ParentInactive(
          'Current time $now is outside parent valid window '
          '[${parent.validFrom}, ${parent.validUntil})');
    }

    if (!Ed25519.verifyDetached(
        _sodium, parent.publicKey, signedBytes, record.parentSignature!)) {
      throw ParentSignatureInvalid(
          'Parent signature invalid for key_id=${_hex(record.parentKeyId!)}');
    }

    _verifyAtDepth(parentBytes, pinned, lookup, depth + 1);
  }

  static String _hex(Uint8List bytes) =>
      bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
}

File: src/main/java/com/erikromson/datawallet/directory/DirectoryRecordVerifier.java
package com.erikromson.datawallet.directory;

import com.erikromson.datawallet.crypto.Ed25519;

import java.time.Clock;
import java.time.Duration;
import java.util.HexFormat;
import java.util.Set;

public final class DirectoryRecordVerifier {

    private static final Duration MAX_AGE = Duration.ofDays(7);
    private static final int MAX_DEPTH = 2;
    private static final Set<String> VALID_RECORD_TYPES = Set.of("verifier", "issuer", "intermediate");
    private static final Set<String> VALID_KEY_USES = Set.of("enc", "auth", "sign");
    private static final Set<String> VALID_STATUSES = Set.of("active", "superseded", "revoked");

    /** Resolves a parent intermediate record by its key_id. Returns the raw signed CBOR bytes. */
    @FunctionalInterface
    public interface ParentLookup {
        byte[] findParentBytes(byte[] parentKeyId);
    }

    private static final ParentLookup NO_PARENT_LOOKUP = parentKeyId -> {
        throw new DirectoryRejection.ParentNotFound(
                "Parent lookup not available for key_id=" + HexFormat.of().formatHex(parentKeyId));
    };

    private final DirectoryRecordCodec codec;
    private final Clock clock;

    public DirectoryRecordVerifier(DirectoryRecordCodec codec, Clock clock) {
        this.codec = codec;
        this.clock = clock;
    }

    public DirectoryRecordVerifier(DirectoryRecordCodec codec) {
        this(codec, Clock.systemUTC());
    }

    /** Verifies a root-signed record. Fails fast if a parent chain is required. */
    public DirectoryRecord verify(byte[] recordBytes, PinnedRoot pinned) {
        return verify(recordBytes, pinned, NO_PARENT_LOOKUP);
    }

    /** Verifies a record, walking the parent chain using [lookup] when needed. */
    public DirectoryRecord verify(byte[] recordBytes, PinnedRoot pinned, ParentLookup lookup) {
        return verifyAtDepth(recordBytes, pinned, lookup, 1);
    }

    private DirectoryRecord verifyAtDepth(byte[] recordBytes, PinnedRoot pinned,
                                           ParentLookup lookup, int depth) {
        DirectoryRecord record = codec.decode(recordBytes);

        if (record.version() != 1) {
            throw new DirectoryRejection.UnsupportedVersion(record.version());
        }
        if (!VALID_RECORD_TYPES.contains(record.recordType())) {
            throw new DirectoryRejection.InvalidRecordType(record.recordType());
        }
        if (!VALID_KEY_USES.contains(record.keyUse())) {
            throw new DirectoryRejection.InvalidKeyUse(record.keyUse());
        }
        if (!VALID_STATUSES.contains(record.status())) {
            throw new DirectoryRejection.InvalidStatus(record.status());
        }

        long nowMs = clock.millis();
        long ageMs = nowMs - record.issuedAt();
        if (ageMs > MAX_AGE.toMillis()) {
            throw new DirectoryRejection.FreshnessExpired(
                    "Record issued_at=" + record.issuedAt() + " is older than 7 days (age=" + ageMs + "ms)");
        }

        byte[] signedBytes = codec.signedBytesOf(recordBytes);

        if (record.parentKeyId() == null) {
            RootQuorum.verify(signedBytes, record.rootSignatures(), record.issuedAt(), pinned);
        } else {
            verifyParentChain(record, signedBytes, pinned, lookup, depth);
        }

        return record;
    }

    private void verifyParentChain(DirectoryRecord record, byte[] signedBytes,
                                    PinnedRoot pinned, ParentLookup lookup, int depth) {
        if (depth >= MAX_DEPTH) {
            throw new DirectoryRejection.ChainTooDeep(
                    "Chain depth exceeds maximum of " + MAX_DEPTH);
        }

        byte[] parentBytes = lookup.findParentBytes(record.parentKeyId());
        DirectoryRecord parent = codec.decode(parentBytes);

        if (!"intermediate".equals(parent.recordType())) {
            throw new DirectoryRejection.ParentNotIntermediate(
                    "Parent record_type=" + parent.recordType() + " is not 'intermediate'");
        }
        if (!"active".equals(parent.status())) {
            throw new DirectoryRejection.ParentInactive(
                    "Parent status=" + parent.status() + " is not 'active'");
        }
        long nowMs = clock.millis();
        if (nowMs < parent.validFrom() || nowMs >= parent.validUntil()) {
            throw new DirectoryRejection.ParentInactive(
                    "Current time " + nowMs + " is outside parent valid window ["
                            + parent.validFrom() + ", " + parent.validUntil() + ")");
        }

        if (!Ed25519.verifyDetached(parent.publicKey(), signedBytes, record.parentSignature())) {
            throw new DirectoryRejection.ParentSignatureInvalid(
                    "Parent signature invalid for key_id="
                            + HexFormat.of().formatHex(record.parentKeyId()));
        }

        verifyAtDepth(parentBytes, pinned, lookup, depth + 1);
    }
}
